src/http/states/readBody/ReadBody.cpp

```diff
@@ -1,146 +1,389 @@
 #include "ReadBody.hpp"
 
+#include "libftpp/utility.hpp"
 #include <client/Client.hpp>
 #include <http/Headers.hpp>
 #include <http/Request.hpp>
 #include <http/StatusCode.hpp>
+#include <http/abnfRules/generalRules.hpp>
+#include <http/abnfRules/headerLineRules.hpp>
+#include <http/abnfRules/http11Rules.hpp>
+#include <http/abnfRules/ruleIds.hpp>
 #include <http/states/prepareResponse/PrepareResponse.hpp>
 #include <libftpp/algorithm.hpp>
+#include <libftpp/memory.hpp>
 #include <libftpp/string.hpp>
 #include <utils/IBuffer.hpp>
+#include <utils/abnfRules/Extractor.hpp>
+#include <utils/abnfRules/Rule.hpp>
+#include <utils/abnfRules/RuleResult.hpp>
+#include <utils/abnfRules/SequenceRule.hpp>
 #include <utils/logger/Logger.hpp>
 #include <utils/state/IState.hpp>
 
 #include <algorithm>
+#include <cstddef>
+#include <cstdlib>
+#include <iostream>
 #include <sstream>
 #include <string>
 
 /* ************************************************************************** */
 // INIT
 
 Logger& ReadBody::_log = Logger::getInstance(LOG_HTTP);
+const std::size_t ReadBody::_readChunkSize;
 
 /* ************************************************************************** */
 // PUBLIC
 
 ReadBody::ReadBody(Client* context)
   : IState<Client>(context)
   , _client(context)
+  , _buffReader()
   , _bodyLength(0)
   , _consumed(0)
-  , _initialized(false)
+  , _chunkedState(ReadChunkInfo)
   , _fixedLengthBody(false)
-  , _chunked(false)
+  , _chunkedBody(false)
   , _done(false)
 {
   _log.info() << "ReadBody\n";
+
+  _determineBodyFraming();
+  _buffReader.init(&_client->getInBuff());
+  _chunkInfoRule().setBufferReader(&_buffReader);
+  _endOfLineRule().setBufferReader(&_buffReader);
+  _fieldLineRule().setBufferReader(&_buffReader);
+
+  _chunkInfoRule().setResultMap(&_results);
+  _endOfLineRule().setResultMap(&_results);
+  _fieldLineRule().setResultMap(&_results);
 }
 
 void ReadBody::run()
-try {
-  if (!_initialized) {
-    _initialized = true;
-    _determineBodyFraming();
-  }
-
+{
   if (_fixedLengthBody) {
     _readFixedLengthBody();
-  } else if (_chunked) {
+  } else if (_chunkedBody) {
     _readChunkedBody();
   }
 
-  if (_done) {
+  if (_done || _client->getResponse().getStatusCode() != StatusCode::Ok) {
     getContext()->getStateHandler().setState<PrepareResponse>();
   }
-} catch (const IBuffer::BufferException& e) {
-  _log.error() << "ReadBody: " << e.what() << '\n';
-  getContext()->getResponse().setStatusCode(StatusCode::InternalServerError);
-  getContext()->getStateHandler().setState<PrepareResponse>();
 }
 
 /* ************************************************************************** */
 // PRIVATE
 
+/**
+ * chunk          = chunk-size [ chunk-ext ] CRLF
+ *                  chunk-data CRLF
+ *
+ * ! only first part
+ */
+SequenceRule& ReadBody::_chunkInfoRule()
+{
+  static SequenceRule rule;
+  static bool init = false;
+  if (!init) {
+    init = true;
+    rule.addRule(chunkInfoRule());
+  }
+  return rule;
```

Comment:
This avoids the extra rule nesting:
```suggestion
  static const ft::shared_ptr<SequenceRule> rule = chunkInfoRule();
  return *rule;
```

The other functions can do the same, just need to change the return types.